src/behavior_generation_lecture_python/graph_search/a_star.py

```diff
@@ -1,160 +1,172 @@
 
+    def total_cost(self) -> float:
+        """
+        Computes the expected total cost to reach the goal node as sum of cost to come and heuristic cost to go.
 
-class ExampleGraph:
-    def __init__(self):
-        HH = GraphNode("HH", 170, 620)
-        H = GraphNode("H", 150, 520)
-        B = GraphNode("B", 330, 540)
-        L = GraphNode("L", 290, 420)
-        F = GraphNode("F", 60, 270)
-        S = GraphNode("S", 80, 120)
-        M = GraphNode("M", 220, 20)
+        :return: The expected total cost to reach the goal node.
+        """
+        return self.cost_to_come + self.heuristic_cost_to_go
 
-        self.nodes = [HH, H, B, L, F, S, M]
 
-        HH.add_connected_to(H)
-        H.add_connected_to(HH)
-        HH.add_connected_to(B)
-        B.add_connected_to(HH)
+def extract_min(node_set: Set[str], node_dict: Dict[str, Node]) -> str:
+    """
+    Extract the node with minimal total cost from a set.
 
-        H.add_connected_to(B)
-        B.add_connected_to(H)
-        H.add_connected_to(L)
-        L.add_connected_to(H)
-        H.add_connected_to(F)
-        F.add_connected_to(H)
+    :param node_set: The set of node names to be considered.
+    :param node_dict: The node dict, containing the node information.
+    :return: The name of the node with minimal total cost.
+    """
+    min_node = min(node_set, key=lambda x: node_dict[x].total_cost())
+    node_set.remove(min_node)
+    return min_node
 
-        B.add_connected_to(L)
-        L.add_connected_to(B)
 
-        L.add_connected_to(S)
-        S.add_connected_to(L)
-        L.add_connected_to(M)
-        M.add_connected_to(L)
+class Graph:
+    def __init__(self, nodes_dict: Dict[str, Node]) -> None:
+        """
+        A graph for A* computation.
 
-        F.add_connected_to(S)
-        S.add_connected_to(F)
+        :param nodes_dict: The dictionary containing the nodes of the graph.
+        """
+        self.nodes_dict = nodes_dict
 
-        S.add_connected_to(M)
-        M.add_connected_to(S)
+        self._end_node = None
 
         fig, ax = plt.subplots()
         self.fig = fig
         self.ax = ax
 
-    def draw(self):
+    def draw_graph(self) -> None:
+        """
+        Draw all nodes and their connections in the graph.
+
+        :return:
+        """
         self.ax.set_xlim([0, 700])
         self.ax.set_ylim([0, 700])
 
-        for node in self.nodes:
-            self.ax.plot(node.x, node.y, marker="o", markersize=6, color="k")
-            self.ax.annotate(node.name, (node.x + 10, node.y + 10))
+        for node in self.nodes_dict.values():
+            self.ax.plot(
+                node.position[0], node.position[1], marker="o", markersize=6, color="k"
+            )
+            self.ax.annotate(node.name, (node.position[0] + 10, node.position[1] + 10))
             for connected_to in node.connected_to:
+                connected_node = self.nodes_dict[connected_to]
                 self.ax.plot(
```

Comment:
Positioning of the distance and the coordinates could be improved relativ to the node. Currently they may overlap with the edges.

Reply:
Also "only" nice to have for me, I'd first merge this, and do the nice-to-have's in a separate PR